### ISSUE TITLE
Unable to install from .deb package registry following manual installation instructions

### ISSUE THREAD

**Issue by Kashemir001** [opentofu/opentofu#1744]
### OpenTofu Version

```shell
not applicable
```

### OpenTofu Configuration Files

```hcl
...tofu config...
```

### Debug Output

`apt update` output:
```
W: GPG error: https://packages.opentofu.org/opentofu/tofu/any any InRelease: The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 70DF59811A8B9109
E: The repository 'https://packages.opentofu.org/opentofu/tofu/any any InRelease' is not signed.
```

### Expected Behavior

`apt` should be able to successfully fetch available packages from registry

### Actual Behavior

Registry verification fails due to corrupted gpg keyring file

### Steps to Reproduce

Follow [step by step](https://opentofu.org/docs/intro/install/deb/#step-by-step-instructions) instruction for installing .deb based package of OpenTofu

### Additional Context

Hello, I tried to install OpenTofu on my Debian 10 machine and went with the manual route, but was unable to fetch package info from the .deb package registry. After some digging, it seems that the problem lays in the following command from the instructions:
```
$ curl -fsSL https://get.opentofu.org/opentofu.gpg | sudo tee /etc/apt/keyrings/opentofu.gpg >/dev/null
```
Trying to read created file:
```
$ gpg --show-keys /etc/apt/keyrings/opentofu.gpg 
gpg: [don't know]: invalid packet (ctb=0a)
gpg: read_block: read error: Invalid packet
gpg: import from '/etc/apt/keyrings/opentofu.gpg' failed: Invalid keyring
$ md5sum /etc/apt/keyrings/opentofu.gpg 
920221492fbe7edd2561d499a36ea47a  /etc/apt/keyrings/opentofu.gpg
$ sudo apt update
...
Err:10 https://packages.opentofu.org/opentofu/tofu/any any InRelease
  The following signatures couldn't be verified because the public key is not available: NO_PUBKEY 70DF59811A8B9109
...
```
`tee` appends `LF` symbol (`0a`) at the end of the file, which makes it corrupted and unusable.

If we use curl's `-o` option to write file content directly:
```
$ sudo curl -fsSL https://get.opentofu.org/opentofu.gpg -o /etc/apt/keyrings/opentofu.gpg
$ gpg --show-keys /etc/apt/keyrings/opentofu.gpg 
pub   rsa4096 2023-11-15 [SC]
      E3E6E43D84CB852EADB0051D0C0AF313E5FD9F80
uid                      OpenTofu (This key is used to sign opentofu providers) <core@opentofu.org>
sub   rsa4096 2023-11-15 [E]
$ md5sum /etc/apt/keyrings/opentofu.gpg 
c4b46ad81f67f5825e21399cfd7f4e06  /etc/apt/keyrings/opentofu.gpg
$ sudo apt update
...
Hit:12 https://packages.opentofu.org/opentofu/tofu/any any InRelease
...
```

Automatic installation script uses either `wget -o` or `curl -o` and does not suffer from this.

By the way, this is .deb packages verification keyring that gets broken, not the repository one, but due to the fact that it is placed first in the `signed-by` option in sources .list file:
```deb [signed-by=/etc/apt/keyrings/opentofu.gpg,/etc/apt/keyrings/opentofu-repo.gpg] ...```
it seems to affect `apt` on the repository fetching step as well. If you flip positions of keyring files in `signed-by`, `apt update` manages to complete successfully, which is not really a solution :)

**Comment by krapgras** [comment]
I'm having the same issue during manual install

**Comment by krapgras** [comment]
Hello @janosdebugs 

I am running this in a docker container on azure devops pipeline but i'm doing the following:

```
  curl -fsSL https://get.opentofu.org/opentofu.gpg -o /usr/share/keyrings/opentofu.gpg && \
  curl -fsSL https://packages.opentofu.org/opentofu/tofu/gpgkey | gpg --dearmor -o /usr/share/keyrings/opentofu-repo.gpg && \
  chmod a+r /usr/share/keyrings/opentofu.gpg /usr/share/keyrings/opentofu-repo.gpg && \
  echo "deb [signed-by=/etc/apt/keyrings/opentofu.gpg,/etc/apt/keyrings/opentofu-repo.gpg] https://packages.opentofu.org/opentofu/tofu/any/ any main" >> /etc/apt/sources.list.d/opentofu.list && \
```

Which throws following error:

![image](https://github.com/opentofu/opentofu/assets/9533723/461f75f1-0d38-4412-8d87-6a246dbcadb8)

However the list file is working because if i change the command to:

`echo "deb [trusted=yes] https://packages.opentofu.org/opentofu/tofu/any/ any main" >> /etc/apt/sources.list.d/opentofu.list && \`

It does work, so clearly it goes wrong in the downloads somewhere.

I have tried with both curl -o and tee commands.

**Comment by Kashemir001** [comment]
@janosdebugs thanks for taking your time to look at it.

I gave it a try in Ubuntu container too, which worked fine, and after some thinking where the extra new line might come from, I found out I had a `~/.curlrc` file with `-w "\n"` in it, set manually years ago. It makes curl print to stdout whatever you pass to it (new line in my case) after it finishes data transfer.
When I tried with `curl ... -o key.gpg`, key content was being redirected to a file, and \n to stdout, so it worked fine for me.
When piped into `tee`, key content and extra new line got mixed together, and that got put into a file.

In short - my case was a client problem. Sorry for bringing this up, I hope it at least might save some other lurkers from a headache.

**Comment by Kashemir001** [comment]
@krapgras looking at the commands you provided, you save key files to `/usr/share/keyrings/`, but use `/etc/apt/keyrings/` path in .list file... Can you verify it is not the source of your problem?

**Comment by krapgras** [comment]
I indeed forgot that last change to /usr/share/keyrings/ in the echo. My bad -_-
However just recommendation for the instructions it would not be bad to change the code to /usr/share/keyrings instead of /etc/apt/keyrings based on the man page info:

> The recommended locations for keyrings are /usr/share/keyrings for keyrings managed by packages, and /etc/apt/keyrings for keyrings managed by the system operator. [manpage](https://manpages.debian.org/unstable/apt/sources.list.5.en.html)

**Comment by skyzyx** [comment]
The `[trusted=yes]` tells apt to auto-trust the repo, regardless of GPG key. If the error is that `InRelease` is not signed, than that error is expected to be ignored when `[trusted=yes]` is set.

**Comment by Kashemir001** [comment]
> The `[trusted=yes]` tells apt to auto-trust the repo, regardless of GPG key. If the error is that `InRelease` is not signed, than that error is expected to be ignored when `[trusted=yes]` is set.

Completely disabling signature validation does not sound like a solution

**Comment by krapgras** [comment]
Hello @skyzyx @Kashemir001,

That is indeed not the solution, that was just a debugging step to see if the package was working regardless of the signature.

We were just looking for more robust way as one issue was the new line that was added by curlrc.

The other suggestion from my side was to move the keyring from /etc/apt/keyrings to /usr/share/keyrings as it is not a system package following the best practice in the man pages.

**Comment by flku-snp** [comment]
Please try saving the attached file [opentofu.sources.txt](https://github.com/user-attachments/files/20205446/opentofu.sources.txt)
to /etc/apt/sources.list.d/opentofu.sources
Does it work as is ?

- I have set a repo name 
- There is an enabled field , to quickly en-/dis-able thr repo
- The public-key is embeded

Still missing is the the source part of the repo, but I wont do that right now, since I want feedback first.
I think this is the direction to take.
